### PR TITLE
Gaussmom

### PR DESCRIPTION
**Pull request by b-biswas**

### CODE REVIEW

File: deep_field_metadetect/gaussmom/gaussmom.py [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+import numpy as np
+from ngmix.moments import MOMENTS_NAME_MAP
+
+from deep_field_metadetect.gaussmom.gaussmom_core import (
+    GaussMomData,
+    GaussMomObs,
+    _eval_gauss2d,
+)
+
+jax.config.update("jax_enable_x64", True)
```
Comment (beckermr) [inline deep_field_metadetect/gaussmom/gaussmom.py]: We should not have this hardcoded here.

File: deep_field_metadetect/gaussmom/gaussmom.py [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+
+    sums, sums_cov, wsum, _ = get_weighted_sums(
+        obs,
+        T=T,
+        maxrad=maxrad,
+    )
+    return get_weighted_moments_stats(obs, sums, sums_cov, wsum)
+
+
+# @jax.jit
```
Comment (beckermr) [inline deep_field_metadetect/gaussmom/gaussmom.py]: Is there a reason the jit decorators are commented out? Do we want to have it by default? I am not sure, but I am leaning towards yes.

File: deep_field_metadetect/gaussmom/gaussmom.py [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+
+    mompars = [0, 0, T / 2, 0, T / 2]
+
+    vmod = gaussmom_obs.v - vcen
+    umod = gaussmom_obs.u - ucen
+
+    var = 1.0 / (gaussmom_obs.pixelwise_wgt)
+    rad2 = umod * umod + vmod * vmod
+
+    circle_mask = jnp.where(jnp.sqrt(rad2) <= maxrad, 1, 0)
```
Comment (beckermr) [inline deep_field_metadetect/gaussmom/gaussmom.py]: Is this masked used in ngmix too? I forgot that detail.

File: deep_field_metadetect/gaussmom/gaussmom.py [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+        mompars, gaussmom_obs.u, gaussmom_obs.v, gaussmom_obs.area
+    )
+
+    wt_noimage = wt_noimage / (wt_norm * gaussmom_obs.area)
+    wt_noimage = circle_mask * wt_noimage
+
+    wdata = wt_noimage * gaussmom_obs.image
+
+    # print(jnp.sum(umod * umod + vmod * vmod))
+
```
Comment (beckermr) [inline deep_field_metadetect/gaussmom/gaussmom.py]: Let's remove commented out code.

File: deep_field_metadetect/gaussmom/gaussmom.py [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+            gaussmom_obs.u,
+            umod * umod - vmod * vmod,
+            2 * vmod * umod,
+            rad2,
+            jnp.ones_like(gaussmom_obs.v),
+        ]
+    )
+
+    wsum = jnp.sum(wt_noimage)
+    # res["npix"] = jnp.sum(circle_mask)
```
Comment (beckermr) [inline deep_field_metadetect/gaussmom/gaussmom.py]: Why is this commented out?

File: deep_field_metadetect/gaussmom/gaussmom.py [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+
+    for i in range(6):
+        for j in range(6):
+            sums_cov = sums_cov.at[i, j].set(jnp.sum(wt_noimage**2 * var * F[i] * F[j]))
+
+    return sums, sums_cov, wsum, wt_norm
+
+
+# @jax.jit
+def get_weighted_moments_stats(obs, sums, sums_cov, sums_norm=None):
```
Comment (beckermr) [inline deep_field_metadetect/gaussmom/gaussmom.py]: We'd like this to be able to be jitted. The current control flow will not enable that.

File: deep_field_metadetect/gaussmom/gaussmom.py [hunk 7]
```diff
@@ -0,0 +1,417 @@
...
+        b == 0, 1e-100, b
+    )  # TODO: This does not raise a value error like ngmix
+
+    rsq = (a / b) ** 2
+
+    var = rsq * (var_a / a**2 + var_b / b**2 - 2 * cov_ab / (a * b))
+    return var
+
+
+def eval_gaussian_moments(gaussmom_obs: GaussMomObs, fwhm: float, maxrad: float):
```
Comment (beckermr) [inline deep_field_metadetect/gaussmom/gaussmom.py]: I don't recall this function in the original ngmix API.

**Review by beckermr — CHANGES_REQUESTED** [review]
I left some comments. One big item is that ideally we'd be able to vmap these functions over the observations in order to run them in parallel. We should add a test for that.

File: deep_field_metadetect/gaussmom/gaussmom.py (hunk 2 above)
Comment (b-biswas) [reply deep_field_metadetect/gaussmom/gaussmom.py]: it was just for debugging, jitting is definitely making it faster

File: deep_field_metadetect/gaussmom/gaussmom.py (hunk 3 above)
Comment (b-biswas) [reply deep_field_metadetect/gaussmom/gaussmom.py]: There is a maxrad variable in ngmix, I am using this mask to mimic that variable

File: deep_field_metadetect/gaussmom/gaussmom.py (hunk 7 above)
Comment (b-biswas) [reply deep_field_metadetect/gaussmom/gaussmom.py]: I am making the API closer to ngmix now

File: deep_field_metadetect/gaussmom/gaussmom.py (hunk 6 above)
Comment (b-biswas) [reply deep_field_metadetect/gaussmom/gaussmom.py]: done!

File: deep_field_metadetect/gaussmom/gaussmom_core.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
+from typing import NamedTuple
+
+import jax
+import jax.numpy as jnp
+import ngmix
+
+
+@jax.tree_util.register_pytree_node_class
+class GaussMomObs(NamedTuple):
```
Comment (beckermr) [inline deep_field_metadetect/gaussmom/gaussmom_core.py:9]: We should use a single observation type for everything. I guess the goal is to move this over to match depp-field mdet later?

File: deep_field_metadetect/gaussmom/tests/test_gaussmom.py [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+#         M22_ind = MOMENTS_NAME_MAP["M22"]
+#         rho4s[i] = res['sums'][M22_ind] / res['sums'][f_ind] / sigma**4
+
+#     rho4_mean = rho4s.mean()
+#     rho4_std = rho4s.std()
+#     print(f'rho4: {rho4_mean:.3g} std: {rho4_std:.3g}')
+#     assert np.abs(rho4_mean - 2) < 1e-5
+
+
+# def test_gaussmom_flags():
```
Comment (beckermr) [inline deep_field_metadetect/gaussmom/tests/test_gaussmom.py]: We should enable a version of this flags tests.

File: deep_field_metadetect/gaussmom/gaussmom_core.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+            self.pixelwise_wgt,
+        ), None
+
+    @classmethod
+    def tree_unflatten(cls, aux_data, children):
+        return cls(*children)
+
+
+@jax.tree_util.register_pytree_node_class
+class GaussMomData(NamedTuple):
```
Comment (beckermr) [inline deep_field_metadetect/gaussmom/gaussmom_core.py:31]: We should be pre-fixing classes and functions with `_` if they are not meant to be used by users.

File: deep_field_metadetect/gaussmom/gaussmom.py [hunk 11]
```diff
@@ -0,0 +1,558 @@
...
+        for i in range(6):
+            for j in range(6):
+                sums_cov = sums_cov.at[i, j].set(
+                    jnp.sum(wt_noimage**2 * var * F[i] * F[j])
+                )
+
+        return sums, sums_cov, wsum, npix
+
+    def tree_flatten(self):
+        children = (self.fwhm, self.with_higher_order)
```
Comment (beckermr) [inline deep_field_metadetect/gaussmom/gaussmom.py:653]: Should this include the `kind` class attribute?

File: deep_field_metadetect/gaussmom/gaussmom.py [hunk 12]
```diff
@@ -0,0 +1,558 @@
...
+    return fwhm / 2.3548200450309493  # sig = fwhm / sqrt(8 * ln 2)
+
+
+@jax.jit
+def fwhm_to_T(fwhm):
+    """
+    convert fwhm to T for a gaussian
+    """
+    sigma = fwhm_to_sigma(fwhm)
+    return 2 * sigma**2
```
Comment (beckermr) [inline deep_field_metadetect/gaussmom/gaussmom.py]: These should go into a separate moments module or similar since they will be useful in other places too.

File: deep_field_metadetect/gaussmom/gaussmom.py [hunk 13]
```diff
@@ -0,0 +1,558 @@
...
+
+
+@jax.jit
+def create_circular_mask(umod, vmod, maxrad):
```
Comment (beckermr) [inline deep_field_metadetect/gaussmom/gaussmom.py]: This function is purely internal I think? If so let's use an `_` at the start.

File: deep_field_metadetect/gaussmom/gaussmom.py [hunk 14]
```diff
@@ -0,0 +1,558 @@
...
+    b = jnp.where(
+        b == 0, 1e-100, b
+    )  # TODO: This does not raise a value error like ngmix
+
+    rsq = (a / b) ** 2
+
+    var = rsq * (var_a / a**2 + var_b / b**2 - 2 * cov_ab / (a * b))
+
+    # var = jnp.where(var>1e100, jnp.inf, var)
+    return var
```
Comment (beckermr) [inline deep_field_metadetect/gaussmom/gaussmom.py]: These should go into a separate module as well. Maybe `stats`?

File: deep_field_metadetect/gaussmom/gaussmom.py [hunk 15]
```diff
@@ -0,0 +1,558 @@
...
+
+    def _measure_moments(
+        self, gaussmom_obs: GaussMomObs, maxrad: float, with_higher_order: bool = False
+    ):
+        res = self.get_weighted_moments(
+            gaussmom_obs=gaussmom_obs,
+            maxrad=maxrad,
+            with_higher_order=with_higher_order,
+        )
+        # e1e2T = mom2e(res[2], res[3], res[4])
```
Comment (beckermr) [inline deep_field_metadetect/gaussmom/gaussmom.py]: Commented code we do not need should be removed.

File: deep_field_metadetect/gaussmom/gaussmom.py [hunk 16]
```diff
@@ -0,0 +1,558 @@
...
+    # Ensure no division by zero
+    b = jnp.where(
+        b == 0, 1e-100, b
+    )  # TODO: This does not raise a value error like ngmix
+
+    rsq = (a / b) ** 2
+
+    var = rsq * (var_a / a**2 + var_b / b**2 - 2 * cov_ab / (a * b))
+
+    # var = jnp.where(var>1e100, jnp.inf, var)
```
Comment (beckermr) [inline deep_field_metadetect/gaussmom/gaussmom.py]: Did you decide against using jnp.inf? What happened here?

File: deep_field_metadetect/gaussmom/gaussmom.py [hunk 17]
```diff
@@ -0,0 +1,558 @@
...
+        e=res_e,
+        e_err=res_e_err,
+        e_cov=res_e_cov,
+        sums_err=res_sums_err,
+        pars=pars,
+    )
+
+    # _add_moments_by_name(res) # shouldn't modify sums or sums_cov
+
+    # TODO: not yet computing the normalized momemnts [Mv, Mu, M1, M2, MT, MF]
```
Comment (beckermr) [inline deep_field_metadetect/gaussmom/gaussmom.py]: Do we need to address these?

Comment (beckermr) [reply deep_field_metadetect/gaussmom/gaussmom.py]: It seems like we should.

**Review by beckermr — CHANGES_REQUESTED** [review]
Looking great! 

The logic in the flags function is pretty intense. Nice job! 

Do we want to enable more detailed tests for that or break that function into smaller pieces for testing?

File: deep_field_metadetect/gaussmom/gaussmom.py (hunk 16 above)
Comment (b-biswas) [reply deep_field_metadetect/gaussmom/gaussmom.py]: This line does not exist in original ngmix. If b==0 ngmix raises a ValueError but to make it jitable, I set b to a very small value. I intoduced this line to make sure that var is retuned inf it is it too large.
I commented it out just to discuss if a better way to go about this

Comment (beckermr) [reply deep_field_metadetect/gaussmom/gaussmom.py]: So my general approach has been to insert NaNs when something catastrophic happens. equinox has APIs for raising errors inside JITed code, but I have not gone down that route. So I'd insert a NaN here for now.

**Review by beckermr — APPROVED** [review]
This pr is fantastic! Well done. Thank you!
